### PR TITLE
Upload csv functionality

### PR DESCRIPTION
**Pull request by jamesjefferies**

### CODE REVIEW

File: app/models/amr_uploaded_reading.rb [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+#
+#  fk_rails_...  (amr_data_feed_config_id => amr_data_feed_configs.id) ON DELETE => cascade
+#
+
+class AmrUploadedReading < ApplicationRecord
+  belongs_to :amr_data_feed_config
+
+  validates_presence_of :file_name
+
+  validate :validate_reading_data, on: :validate_reading_data
```
Comment (jamesjefferies) [inline app/models/amr_uploaded_reading.rb]: Only save it if actually valid? You can still show the errors etc...

Comment (jamesjefferies) [reply app/models/amr_uploaded_reading.rb]: Actually let's have a ReadingData activemodel object which we can then use anywhere....

File: app/controllers/admin/amr_data_feed_readings_controller.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
+module Admin
+  class AmrDataFeedReadingsController < AdminController
+    def create
```
Comment (jamesjefferies) [inline app/controllers/admin/amr_data_feed_readings_controller.rb]: This does the actual insert of the data

File: app/services/amr/csv_parser.rb [hunk 3]
```diff
@@ -2,13 +2,13 @@
...
-    def initialize(config, file_name)
-      @file_name = file_name
+    def initialize(config, path_and_file_name)
+      @path_and_file_name = path_and_file_name
       @config = config
     end
 
     def perform
-      CSV.read("#{@config.local_bucket_path}/#{@file_name}", col_sep: @config.column_separator, row_sep: :auto)
+      CSV.read(@path_and_file_name, col_sep: @config.column_separator, row_sep: :auto)
```
Comment (jamesjefferies) [inline app/services/amr/csv_parser.rb]: This allows us to parse temporary files which are not in the correct local bucket path

File: app/services/amr/csv_to_amr_reading_data.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
+module Amr
+  class CsvToAmrReadingData
```
Comment (jamesjefferies) [inline app/services/amr/csv_to_amr_reading_data.rb]: This replaces CSV to Reading Hash - now returns object

File: app/controllers/admin/amr_data_feed_readings_controller.rb [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+      amr_upload_reading = AmrUploadedReading.find(params[:amr_uploaded_reading_id])
+
+      amr_data_feed_import_log = AmrDataFeedImportLog.create(
+        amr_data_feed_config_id: amr_upload_reading.amr_data_feed_config_id,
+        file_name: amr_upload_reading.file_name,
+        import_time: DateTime.now.utc
+                                                            )
+      @upserted_record_count = Amr::DataFeedUpserter.new(amr_upload_reading.reading_data, amr_data_feed_import_log.id).perform
```
Comment (jamesalmond) [inline app/controllers/admin/amr_data_feed_readings_controller.rb]: Do we need to update the log with the count?

File: app/services/amr/csv_parser_and_upserter.rb [hunk 6]
```diff
@@ -13,24 +13,17 @@ def perform
 
       amr_data_feed_import_log = AmrDataFeedImportLog.create(amr_data_feed_config_id: @config.id, file_name: @file_name, import_time: DateTime.now.utc)
 
-      @upserted_record_count = DataFeedUpserter.new(get_unique_array_of_readings, amr_data_feed_import_log.id).perform
+      amr_reading_data = CsvToAmrReadingData.new(@config, "#{@config.local_bucket_path}/#{@file_name}").perform
```
Comment (jamesalmond) [inline app/services/amr/csv_parser_and_upserter.rb]: Should we check the validity here?

File: app/controllers/admin/amr_data_feed_readings_controller.rb (hunk 5 above)
Comment (jamesjefferies) [reply app/controllers/admin/amr_data_feed_readings_controller.rb]: Yes! Thanks for the reminder

File: app/services/amr/csv_parser_and_upserter.rb (hunk 6 above)
Comment (jamesjefferies) [reply app/services/amr/csv_parser_and_upserter.rb]: Yes, we now do check the validity and update the Amr Import Log if there's a problem

**Pull request merged**